precompiles/staking/staking.go

```diff
@@ -0,0 +1,428 @@
+package staking
+
+import (
+	"fmt"
+	"math/big"
+
+	"cosmossdk.io/math"
+	"github.com/cosmos/cosmos-sdk/codec"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	stakingkeeper "github.com/cosmos/cosmos-sdk/x/staking/keeper"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+
+	ptypes "github.com/zeta-chain/node/precompiles/types"
+)
+
+const (
+	// write
+	StakeMethodName         = "stake"
+	UnstakeMethodName       = "unstake"
+	TransferStakeMethodName = "transferStake"
+
+	// read
+	GetAllValidatorsMethodName = "getAllValidators"
+	GetSharesMethodName        = "getShares"
+)
+
+var (
+	ABI                 abi.ABI
+	ContractAddress     = common.HexToAddress("0x0000000000000000000000000000000000000066")
+	GasRequiredByMethod = map[[4]byte]uint64{}
+	ViewMethod          = map[[4]byte]bool{}
+)
+
+func init() {
+	initABI()
+}
+
+func initABI() {
+	if err := ABI.UnmarshalJSON([]byte(IStakingMetaData.ABI)); err != nil {
+		panic(err)
+	}
+
+	GasRequiredByMethod = map[[4]byte]uint64{}
+	for methodName := range ABI.Methods {
+		var methodID [4]byte
+		copy(methodID[:], ABI.Methods[methodName].ID[:4])
+		switch methodName {
+		// TODO: just temporary flat values, double check these flat values
```

Comment:
Can we create an issue? It can group all the part we need to work on for the gas consumption for precompiles